### PR TITLE
extend core proxy options and handling

### PR DESCRIPTION
**Pull request by Diapolo**
- rework the proxy handling in init to cover more cases and work more
  thoroughly
- extend RPC call getnetwork info to see if IPv6 and Tor proxy is the
  default proxy

See #2575 for discussion!

### CODE REVIEW

**Comment by fanquake** [comment]
Needs rebase

**Comment by Diapolo** [comment]
Rebased, needs review / testing.

**Comment by laanwj** [comment]
To be clear what this is blocked on: We need (RPC) tests for the proxy functionality, before we can merge changes like this.

**Comment by Diapolo** [comment]
Rebased to include the recent ``-proxyrandomize`` change. Also as a proxy test was added, this should be less controversial if tests pass I guess :).

File: src/rpcnet.cpp [hunk 1]
```diff
@@ -365,14 +365,20 @@ static Array GetNetworksInfo()
...
         GetProxy(network, proxy);
+
+        Object obj;
         obj.push_back(Pair("name", GetNetworkName(network)));
         obj.push_back(Pair("limited", IsLimited(network)));
         obj.push_back(Pair("reachable", IsReachable(network)));
         obj.push_back(Pair("proxy", proxy.IsValid() ? proxy.proxy.ToStringIPPort() : string()));
         obj.push_back(Pair("proxy_randomize_credentials", proxy.randomize_credentials));
+        if (network != NET_IPV4)
+            obj.push_back(Pair("proxy-from-default", proxy.proxy == proxyDefault.proxy));
```
Comment (laanwj) [inline src/rpcnet.cpp]: Let's remove `proxy-from-default`, it's not useful information and will always be true anyway according to the above code.

Comment (Diapolo) [inline src/rpcnet.cpp]: I'll remove it :).

**Comment by Diapolo** [comment]
@laanwj Removed changes to rpcnet.cpp.

**Comment by Diapolo** [comment]
@laanwj This is passing proxy tests and I removed what you didn't like. Anything I can do to get this forward finally?

**Comment by laanwj** [comment]
I'm terribly sorry sorry for holding this up so long, but I'm not convinced that the new logic is any better.

The logical flow here sounds to me:

- Parse default `-proxy`, apply to *all* nets, as well as the name proxy
- Parse onion-specific proxy `-onion`, apply to `NET_TOR`. This overrides the default set in the first step, if any
*All* errors during parsing of either setting should be fatal, to avoid that the user is running with different settings than they expect.

This is nearer to the original code than yours. I certainly think the code could be simplified and made more clear, what do you think about:
```c++
    bool proxyRandomize = GetBoolArg("-proxyrandomize", true);
    // -proxy sets a proxy for all outgoing network traffic
    // -noproxy (or -proxy=0) as well as the empty string can be used to not set a proxy, this is the default
    std::string proxyArg = GetArg("-proxy", "");
    if (proxyArg != "" && proxyArg != "0") {
        proxyType addrProxy = proxyType(CService(proxyArg, 9050), proxyRandomize);
        if (!addrProxy.IsValid())
            return InitError(strprintf(_("Invalid -proxy address: '%s'"), proxyArg));

        SetProxy(NET_IPV4, addrProxy);
        SetProxy(NET_IPV6, addrProxy);
        SetProxy(NET_TOR, addrProxy);
        SetNameProxy(addrProxy);
        SetReachable(NET_TOR); // by default, -proxy sets onion as reachable, unless -noonion later
    }

    // -onion can override normal proxy for .onion addresses
    // -noonion (or -onion=0) disables connecting to .onion entirely
    // An empty string is used to just not override the onion proxy
    std::string onionArg = GetArg("-onion", "");
    if (onionArg != "") {
        if (onionArg == "0") { // Handle -noonion/-onion=0
            SetReachable(NET_TOR, false); // set onions as unreachable
        } else {
            proxyType addrOnion;
            addrOnion = proxyType(CService(onionArg, 9050), proxyRandomize);
            if (!addrOnion.IsValid())
                return InitError(strprintf(_("Invalid -onion address: '%s'"), onionArg));
            SetProxy(NET_TOR, addrOnion);
            SetReachable(NET_TOR);
        }
    }
```

This provides a straightforward flow, gets rid of `.count()` (which is what you needed for the GUI, as it makes it possible to override an earlier provided proxy option to nothing), as well as comments the different cases.

**Comment by laanwj** [comment]
Closing in favor of #6272

**Pull request closed without merging**